File: agent-app/src/test/kotlin/com/riwonace/agent/sql/SchemaPromptFormatterTest.kt
package com.riwonace.agent.sql

import com.fasterxml.jackson.databind.ObjectMapper
import kotlin.test.Test
import kotlin.test.assertContains
import kotlin.test.assertEquals
import kotlin.test.assertFalse

class SchemaPromptFormatterTest {

    private val formatter = SchemaPromptFormatter(ObjectMapper())

    @Test
    fun `테이블 외래키 값 힌트의 의미 영역을 분리한다`() {
        val raw = """
            {
              "tables": {
                "clients": ["id (integer)", "region (character varying)"],
                "sales": ["client_id (integer)", "amount (integer)"]
              },
              "foreignKeys": ["sales.client_id -> clients.id"],
              "valueHints": {"clients.region": ["서울", "경기"]}
            }
        """.trimIndent()

        val result = formatter.format(raw)

        assertContains(result, "TABLE clients (id (integer), region (character varying))")
        assertContains(result, "sales.client_id -> clients.id")
        assertContains(result, "clients.region = ['서울', '경기']")
        assertContains(result, "SQL 식별자가 아닌 실제 데이터 값")
        assertFalse(result.contains("TABLE valueHints"))
    }

    @Test
    fun `작은따옴표가 있는 실제 값은 SQL 리터럴 규칙으로 이스케이프한다`() {
        val raw = """{"tables":{"clients":["name (varchar)"]},"valueHints":{"clients.name":["O'Reilly"]}}"""

        assertContains(formatter.format(raw), "'O''Reilly'")
    }

    @Test
    fun `다른 MCP 구현의 비 JSON 응답은 원문으로 폴백한다`() {
        val raw = "TABLE clients(id integer)"

        assertEquals(raw, formatter.format(raw))
    }
}

File: agent-app-go/internal/sql/schema_prompt_formatter_test.go
package sql

import (
	"strings"
	"testing"
)

func TestSchemaPromptFormatter_SeparatesTableForeignKeyValueHintSemanticAreas(t *testing.T) {
	f := &SchemaPromptFormatter{}
	raw := `{
		"tables": {
			"clients": ["id (integer)", "region (character varying)"],
			"sales": ["client_id (integer)", "amount (integer)"]
		},
		"foreignKeys": ["sales.client_id -> clients.id"],
		"valueHints": {"clients.region": ["서울", "경기"]}
	}`

	result := f.Format(raw)

	if !strings.Contains(result, "TABLE clients (id (integer), region (character varying))") {
		t.Fatalf("got %q", result)
	}
	if !strings.Contains(result, "sales.client_id -> clients.id") {
		t.Fatalf("got %q", result)
	}
	if !strings.Contains(result, "clients.region = ['서울', '경기']") {
		t.Fatalf("got %q", result)
	}
	if !strings.Contains(result, "SQL 식별자가 아닌 실제 데이터 값") {
		t.Fatalf("got %q", result)
	}
	if strings.Contains(result, "TABLE valueHints") {
		t.Fatalf("got %q", result)
	}
}

func TestSchemaPromptFormatter_SingleQuoteValuesEscapedAsSqlLiterals(t *testing.T) {
	f := &SchemaPromptFormatter{}
	raw := `{"tables":{"clients":["name (varchar)"]},"valueHints":{"clients.name":["O'Reilly"]}}`
	result := f.Format(raw)
	if !strings.Contains(result, "'O''Reilly'") {
		t.Fatalf("got %q", result)
	}
}

func TestSchemaPromptFormatter_NonJsonResponseFromOtherMcpImplFallsBackToRaw(t *testing.T) {
	f := &SchemaPromptFormatter{}
	raw := "TABLE clients(id integer)"
	if got := f.Format(raw); got != raw {
		t.Fatalf("got %q", got)
	}
}
